instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/tracer/BaseTracer.java

```diff
@@ -255,25 +247,35 @@ public void endExceptionally(Context context, Throwable throwable) {
    * end of the span's execution.
    *
    * @param endTimeNanos Explicit nanoseconds timestamp from the epoch.
+   * @see #onException(Context, Throwable)
+   * @see #end(Context)
    */
   public void endExceptionally(Context context, Throwable throwable, long endTimeNanos) {
-    Span span = Span.fromContext(context);
-    span.setStatus(StatusCode.ERROR);
-    onError(span, unwrapThrowable(throwable));
+    onException(context, throwable);
     end(context, endTimeNanos);
   }
 
-  protected void onError(Span span, Throwable throwable) {
-    addThrowable(span, throwable);
+  /**
+   * Records the {@code throwable} in the span stored in the passed {@code context} and sets the
+   * span's status to {@link StatusCode#ERROR}. The throwable is unwrapped ({@link
+   * #unwrapThrowable(Throwable)}) before being added to the span.
+   */
+  protected void onException(Context context, Throwable throwable) {
+    Span span = Span.fromContext(context);
+    span.setStatus(StatusCode.ERROR);
+    span.recordException(unwrapThrowable(throwable));
   }
 
+  /**
+   * Extracts the actual cause by unwrapping passed {@code throwable} from known wrapper exceptions,
+   * e.g {@link ExecutionException}.
+   */
   protected Throwable unwrapThrowable(Throwable throwable) {
-    return throwable instanceof ExecutionException ? throwable.getCause() : throwable;
-  }
-
-  // TODO: call onError instead and make this private
-  public void addThrowable(Span span, Throwable throwable) {
-    span.recordException(throwable);
+    if (throwable.getCause() != null
+        && (throwable instanceof ExecutionException || throwable instanceof CompletionException)) {
```

Comment:
It's tempting to add `InvocationTargetException` and `UndeclaredThrowableException` here too. WDYT?

Reply:
Do we need to do recursive unwrapping, as in `PlayTracer`?

Reply:
Done